File: javascript/19-tangled-web/src/config_loader.js
import fs from 'fs';

class ConfigLoader {
  constructor() {}

  load_config() {
    const defaults = { retries: 3, theme: 'standard', discount: 0 };
    let local = {};
    try {
      const path = process.env.APP_CONFIG_PATH || '/tmp/config.json';
      let raw = fs.readFileSync(path);
      local = JSON.parse(raw);
    } catch (e) {
      local = {};
    }

    const envConfig = {};
    if (process.env.APP_THEME) {
      envConfig.theme = process.env.APP_THEME;
    }
    if (process.env.APP_RETRIES) {
      envConfig.retries = Number.parseInt(process.env.APP_RETRIES, 10);
    }

    let seasonal = {};
    let month = Number.parseInt(process.env.APP_CURRENT_MONTH || `${new Date().getMonth() + 1}`, 10);
    if (month === 12 || month === 1 || month === 2) {
      seasonal = { theme: 'winter', discount: 0.1 };
    } else if (month >= 6 && month <= 8) {
      seasonal = { theme: 'summer', discount: 0.05 };
    }

    return { ...defaults, ...local, ...envConfig, ...seasonal };
  }
}

export { ConfigLoader };

File: ruby/19-tangled-web/src/config_loader.rb
# frozen_string_literal: true

require 'json'
class ConfigLoader
  def load_config
    defaults = { retries: 3, theme: 'standard', discount: 0 }
    begin
      path = ENV['APP_CONFIG_PATH'] || '/tmp/config.json'
      local = JSON.parse(File.read(path), symbolize_names: true)
    rescue StandardError
      local = {}
    end
    env_config = {}
    env_config[:theme] = ENV['APP_THEME'] if ENV['APP_THEME']
    env_config[:retries] = ENV['APP_RETRIES'].to_i if ENV['APP_RETRIES']
    seasonal = {}
    month = (ENV['APP_CURRENT_MONTH'] || Time.now.month).to_i
    seasonal = { theme: 'winter', discount: 0.1 } if [12, 1, 2].include?(month)
    seasonal = { theme: 'summer', discount: 0.05 } if month >= 6 && month <= 8
    defaults.merge(local).merge(env_config).merge(seasonal)
  end
end
